### ISSUE TITLE
[HOLD for payment 2024-08-09] [$250] Workspace switcher - WS switcher and LHN are not reset to Expensify after leaving workspace

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#43981]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.85-4
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
Precondition:
- User is invited to a workspace.

1. Go to staging.new.expensify.com
2. Open workspace switcher.
3. Select the invited workspace.
4. Go to Account settings > Workspaces.
5. Click 3-dot menu on the invited workspace > Leave.
6. Return to LHN.

## Expected Result:
- Workspace switcher will switch to Expensify.
- LHN will show all the chats because Expensify is selected.

## Actual Result:
- Workspace switcher displays Expensify, but when opening the workspace list, Expensify is not selected.
- LHN still shows the filtered chats.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/728392e4-6aaf-4edb-9c7a-c3889e629419

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018dbe15d720ffa01b</li>
        <li>Upwork Job ID: 1804259959268757378</li>
        <li>Last Price Increase: 2024-06-21</li>
<li>Automatic offers: </li>
<ul>
<li>truph01 | Contributor | 103068211</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by izarutskaya** [comment]
We think this issue might be related to the #collect project. #vip-vsb

**Comment by tsa321** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Workspace switcher - WS switcher and LHN are not reset to Expensify after leaving workspace

### What is the root cause of that problem?

In Policy.leaveWorkspace we don't have logic to check current workspace is leaved workspace, then navigate to all expensify chats.

### What changes do you think we should make in order to solve the problem?

In Policy.leaveWorkspace :
We must check if current route / path is leaved workspace then navigate to all expensify workspace.

**Comment by truph01** [comment]
## Proposal 
 
### Please re-state the problem that we are trying to solve in this issue.

Workspace switcher displays Expensify, but when opening the workspace list, Expensify is not selected.
LHN still shows the filtered chats.


### What is the root cause of that problem?
In deleteWorkspace function 

https://github.com/Expensify/App/blob/d938880fbb05077fc4ac634f6d2bc7d22d6072d3/src/libs/actions/Policy/Policy.ts#L335-L340

We have logic to update the lastAccessedWorkspacePolicyID, but we missed it in the leaveWorkspace function

### What changes do you think we should make in order to solve the problem?


### What alternative solutions did you explore? (Optional)
https://github.com/Expensify/App/blob/d938880fbb05077fc4ac634f6d2bc7d22d6072d3/src/libs/actions/Policy/Policy.ts#L716

here, we also should add 

```
if (policyID === lastAccessedWorkspacePolicyID) {
        updateLastAccessedWorkspace(undefined);
    }
```

**Comment by s77rt** [comment]
@tsa321

> In Policy.leaveWorkspace we don't have logic to check current workspace is leaved workspace, then navigate to all expensify chats

This is not the root cause. I don't see why such logic should be a part of the `leaveWorkspace` call. Also this will not fix the problem if the admin removed you (as a member) from the workspace

**Comment by s77rt** [comment]
@truph01 Same note as above ^ Adding the check to the `leaveWorkspace` will only fix the problem if you leave the workspace but what about if another admin removes you from that workspace?

**Comment by tsa321** [comment]
@s77rt How about we make a callback in policy collection, then check if policy is deleted or user is not member of the currently accessed policy anymore, then updateLastAccessedWorkspace(undefined); 

## Proposal

### [updated](https://github.com/Expensify/App/issues/43981#issuecomment-2178315431)

**Comment by s77rt** [comment]
@tsa321 I'd prefer not to relay on Onyx callbacks. Can we go with the same approach as reports? We have a `useEffect` on `ReportScreen` that handles report deletion

**Comment by truph01** [comment]
### [Proposal updated](https://github.com/Expensify/App/issues/43981#issuecomment-2178328769)

- @s77rt I updated my main solution based on your suggestion [above](https://github.com/Expensify/App/issues/43981#issuecomment-2184935447).

**Comment by s77rt** [comment]
@truph01 I noticed that If I select a workspace that I created and then delete it, the bug does not occur. Can you investigate that and why the bug only occurs on invited workspaces? That should help us find a better solution

**Comment by truph01** [comment]
@s77rt 
- When deleting the workspace, we call:
https://github.com/Expensify/App/blob/060160e892b9bfaddab75614817b4b906a2c7b3f/src/pages/workspace/WorkspacesListPage.tsx#L141-L144
it is the same as what we do when switching to "Expensify" workspace. So the bug does not occur when we delete our workspaces that we are owner.

**Comment by truph01** [comment]
### [Update comment](https://github.com/Expensify/App/issues/43981#issuecomment-2191945061)

**Comment by s77rt** [comment]
@truph01 Ah right. In your proposal we are not calling `setActiveWorkspaceID` though but `updateLastAccessedWorkspace`. Won't this make the workspace switcher page has not selected option?

**Comment by truph01** [comment]
@s77rt Yes, ưe cannot call ```setActiveWorkspaceID``` because we already call it in:
https://github.com/Expensify/App/blob/ab2426c91c2bf31ed6d9f02952ee0ab026019169/src/libs/Navigation/NavigationRoot.tsx#L127-L141

**Comment by s77rt** [comment]
@truph01 Makes sense! Overall the proposal looks good to me

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/43981#issuecomment-2178328769)

**Comment by johncschuster** [comment]
@s77rt are you assigning this to @truph01?

**Comment by johncschuster** [comment]
@s77rt are you assigning this to @truph01?

**Comment by s77rt** [comment]
@johncschuster Yes! Let me know if you have any concerns regarding the proposal

**Comment by s77rt** [comment]
Not overdue. Waiting on proposal review and assignment

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: More like a missed case https://github.com/Expensify/App/pull/33280
- The offending PR has been commented on: https://github.com/Expensify/App/pull/33280#discussion_r1703438679
- A discussion in #expensify-bugs has been started: Not needed, one of a kind bug
- Determine if we should create a regression test for this bug: No, low priority

**Comment by johncschuster** [comment]
Acknowledged! I'll issue payment after the regression threshold is cleared.

**Comment by johncschuster** [comment]
# Payment Summary
Contributor: @truph01 paid $250 via [Upwork](https://www.upwork.com/nx/wm/offer/103068211)
Contributor+: @s77rt  owed $250 via NewDot

**Comment by johncschuster** [comment]
Payment has been issued via Upwork to @truph01. Go ahead and request payment on ND, @s77rt! Thank you both for your contributions! 🎉

**Comment by JmillsExpensify** [comment]
$250 approved for @s77rt
